File: SPCode/CS/Security/Support/LoginControl.cs
using System.Text.RegularExpressions;
using System.Diagnostics;
using System;
using System.Drawing.Drawing2D;
using System.Windows.Forms;
using System.Collections;
using System.Drawing;
using Microsoft.VisualBasic;
using System.Data.SqlClient;
using System.Data;
using System.Collections.Generic;
using WeifenLuo.WinFormsUI;
using Microsoft.Win32;
using WeifenLuo;
using System.Threading;
using System.Security.Principal;
using System.Reflection;
using System.ComponentModel;

//Questions? Comments? go to
//http://www.idesign.net

namespace SoftLogik.Win
{
	#region Using directives
	
	
	#endregion
	
	namespace Security
	{
		[DefaultEvent("LoginEvent"), ToolboxBitmap(typeof(SPLoginControl), "LoginControl.bmp")]public abstract partial class SPLoginControl
		{
			
			
			private string m_ApplicationName = string.Empty;
			private bool m_CacheRoles = false;
			
			public delegate void LoginEventEventHandler(object Of);
			private LoginEventEventHandler LoginEventEvent;
			
			public event LoginEventEventHandler LoginEvent
			{
				add
				{
					LoginEventEvent = (LoginEventEventHandler) System.Delegate.Combine(LoginEventEvent, value);
				}
				remove
				{
					LoginEventEvent = (LoginEventEventHandler) System.Delegate.Remove(LoginEventEvent, value);
				}
			}
			
			
			[Category("Credentials")]public bool CacheRoles
			{
				get
				{
					return m_CacheRoles;
				}
				set
				{
					m_CacheRoles = value;
				}
			}
			[Category("Credentials")]public string ApplicationName
			{
				get
				{
					return m_ApplicationName;
				}
				
				set
				{
					m_ApplicationName = value;
				}
			}
			
			private string GetAppName()
			{
				if (ApplicationName != string.Empty)
				{
					return ApplicationName;
				}
				System.Reflection.Assembly clientAssembly = System.Reflection.Assembly.GetEntryAssembly();
				AssemblyName assemblyName = clientAssembly.GetName();
				return assemblyName.Name;
			}
			public static void Logout()
			{
				CustomPrincipal customPrincipal = Thread.CurrentPrincipal as customPrincipal;
				if (customPrincipal != null)
				{
					customPrincipal.Detach();
				}
			}
			public static bool IsLoggedIn
			{
				get
				{
					return Thread.CurrentPrincipal is CustomPrincipal;
				}
			}
			
			virtual public void OnLogin(object sender, EventArgs e)
			{
				string userName = m_UserNameBox.Text;
				string password = m_PasswordBox.Text;
				
				if (userName == string.Empty)
				{
					m_ErrorProvider.SetError(m_UserNameBox, "Please Enter User Name");
					return;
				}
				else
				{
					m_ErrorProvider.SetError(m_UserNameBox, string.Empty);
				}
				if (password == string.Empty)
				{
					m_ErrorProvider.SetError(m_PasswordBox, "Please Enter Password");
					return;
				}
				else
				{
					m_ErrorProvider.SetError(m_PasswordBox, string.Empty);
				}
				string applicationName = GetAppName();
				IUserManager userManager = GetUserManager();
				
				bool authenticated = userManager.Authenticate(applicationName, userName, password);
				if (authenticated)
				{
					IIdentity identity = new GenericIdentity(userName);
					CustomPrincipal.Attach(identity, applicationName, userManager, CacheRoles);
				}
				LoginEventArgs loginEventArgs = new LoginEventArgs(authenticated);
				
				if (LoginEventEvent != null)
					LoginEventEvent(this, loginEventArgs);
			}
			
			protected abstract IUserManager GetUserManager();
			
		}
		
		
		
	}
}

File: SPCode/Security/Support/LoginControl.vb
'Questions? Comments? go to 
'http://www.idesign.net

#Region "Using directives"

Imports System
Imports System.Threading
Imports System.Security.Principal
Imports System.Diagnostics
Imports System.Reflection
Imports System.Drawing
Imports System.Windows.Forms
Imports System.ComponentModel

#End Region

Namespace Security
    <DefaultEvent("LoginEvent"), ToolboxBitmap(GetType(SPLoginControl), "LoginControl.bmp")> _
    Public MustInherit Class SPLoginControl

        Private m_ApplicationName As String = String.Empty
        Private m_CacheRoles As Boolean = False

        Public Event LoginEvent As EventHandler(Of LoginEventArgs)

        <Category("Credentials")> _
        Public Property CacheRoles() As Boolean
            Get
                Return m_CacheRoles
            End Get
            Set(ByVal value As Boolean)
                m_CacheRoles = value
            End Set
        End Property
        <Category("Credentials")> _
        Public Property ApplicationName() As String
            Get
                Return m_ApplicationName
            End Get

            Set(ByVal value As String)
                m_ApplicationName = value
            End Set
        End Property

        Private Function GetAppName() As String
            If ApplicationName <> String.Empty Then
                Return ApplicationName
            End If
            Dim clientAssembly As System.Reflection.Assembly = System.Reflection.Assembly.GetEntryAssembly()
            Dim assemblyName As AssemblyName = clientAssembly.GetName()
            Return assemblyName.Name
        End Function
        Public Shared Sub Logout()
            Dim customPrincipal As CustomPrincipal = TryCast(Thread.CurrentPrincipal, CustomPrincipal)
            If Not customPrincipal Is Nothing Then
                customPrincipal.Detach()
            End If
        End Sub
        Public Shared ReadOnly Property IsLoggedIn() As Boolean
            Get
                Return TypeOf Thread.CurrentPrincipal Is CustomPrincipal
            End Get
        End Property

        Protected Overridable Sub OnLogin(ByVal sender As Object, ByVal e As EventArgs) Handles m_LogInButton.Click
            Dim userName As String = m_UserNameBox.Text
            Dim password As String = m_PasswordBox.Text

            If userName = String.Empty Then
                m_ErrorProvider.SetError(m_UserNameBox, "Please Enter User Name")
                Return
            Else
                m_ErrorProvider.SetError(m_UserNameBox, String.Empty)
            End If
            If password = String.Empty Then
                m_ErrorProvider.SetError(m_PasswordBox, "Please Enter Password")
                Return
            Else
                m_ErrorProvider.SetError(m_PasswordBox, String.Empty)
            End If
            Dim applicationName As String = GetAppName()
            Dim userManager As IUserManager = GetUserManager()

            Dim authenticated As Boolean = userManager.Authenticate(applicationName, userName, password)
            If authenticated Then
                Dim identity As IIdentity = New GenericIdentity(userName)
                CustomPrincipal.Attach(identity, applicationName, userManager, CacheRoles)
            End If
            Dim loginEventArgs As LoginEventArgs = New LoginEventArgs(authenticated)

            RaiseEvent LoginEvent(Me, loginEventArgs)
        End Sub

        Protected MustOverride Function GetUserManager() As IUserManager

    End Class



End Namespace